### PR TITLE
Refactor error handling

### PR DESCRIPTION
**Pull request by Patrik-Stas**
Is addressing https://github.com/hyperledger/aries-vcx/issues/582

changes:
- Removed dead code
- Previously there was `VcxError` defined in `aries-vcx`, this was used by both `aries-vcx` and `libvcx`. Because `libvcx` maps needs to map error types to u32, encoding them for FFI interface, this was part of `aries-vcx` - but obivously shouldn't. So to fix this:
   - `aries-vcx` still defined `VcxError`, but stripped of to minimum needed - mainly removed `u32` mapping etc.
   -  in `libvcx` was created module `errors` which defines `LibvcxError` which includes `u32` mappings
   - to bridge `VcxError` with `LibvcxError`, there's mapping `impl From<VcxErrorKind> for LibvcxErrorKind` in  `mapping_from_ariesvcx.rs` 
- In general, the `errors.rs` files has also been broken down into smaller pieces, typically into files such as `errors/mapping_from_<crate>` which maps errors from `<crate>` to domain errors of the given crate (such as `mapping_from_ariesvcx.rs` mentioned earlier)
- All our crates now consistently contain module `src/errors` 
- Some methods in `api_c` contained more logic than they should - the logic has been moved to `api_handle` layer instead (for example the entire file `api_handle/wallet.rs` has been extracted from code in `api_c` and shifted layer down, to `wallet.rs`). This is good step to separate `api_handle` and `api_c` into separate crates https://github.com/hyperledger/aries-vcx/issues/700
- libvcx: change the way u32 error codes and `LibvcxErrorKind` are associated - there's static vector of `(LibvcxErrorKind, u32)` tuples. This way, adding new error code is minimal effort, as this would be the only single place where mapping is handled - as opposed to previously defined `Kind->u32` and `u32->Kind` mapping declarations, which was quite error prone.
- Additionally renamed all error and error kinds, so there is:
   - libvcx: `ErrorLibvcx`, `ErrorLibvcx`
   - aries-vcx: `ErrorAriesVcx`, `ErrorKindAriesVcx`
   - messages: `ErrorMessages`, `ErrorKindMessages`
   - agency client: `ErrorAgencyClient`, `ErrorKindAgencyClient`

### CODE REVIEW

**Comment by Patrik-Stas** [comment]
As the previous PR got merged and squashed at merge, this created bunch of merge conflicts, I'll probably squash this entire PR after review as well, to get around it.

**Comment by gmulhearn** [comment]
Not sure if i'll have time for review tonight sorry - hopefully tomorrow. But just had an initial look - was wondering, what inspired the change from `XYZError` to `ErrorXYZ`? 

My understanding was that `XYZError` was the norm in rust crates, (if not just calling it `Error` (i.e. `aries_vcx::Error`))

**Comment by Patrik-Stas** [comment]
@gmulhearn good point about errors, I didn't realize there was this convention but looking at bunch of crates, you are right. I'll keep the original naming style

File: aries_vcx/src/errors/mod.rs [hunk 1]
```diff
@@ -0,0 +1,5 @@
+mod mapping_vdrtools;
+mod mapping_agency_client;
+mod mapping_messages;
+mod mapping_others;
+pub mod error;
```
Comment (gmulhearn) [inline aries_vcx/src/errors/mod.rs:5]: there are some more error mappings within the credx_anoncreds and indy_vdr_ledger - maybe they should be moved into one of these mapping modules?

File: libvcx/src/api_lib/api_c/issuer_credential.rs [hunk 2]
```diff
@@ -550,51 +550,13 @@ pub extern "C" fn vcx_issuer_send_credential(
...
-                );
-                cb(command_handle, err.into(), ptr::null_mut());
-            }
-        };
-
-        Ok(())
-    });
-
-    error::SUCCESS.code_num
+    error!("vcx_issuer_get_credential_msg >>> not implemented");
```
Comment (gmulhearn) [inline libvcx/src/api_lib/api_c/issuer_credential.rs:558]: just double checking - is this right? was the previous implementation not correct?

File: libvcx/src/api_lib/api_handle/wallet.rs [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    map_ariesvcx_result(wallet.verify(vk, msg, signature).await)
+}
+
+pub async fn replace_did_keys_start(did: &str) -> LibvcxResult<String> {
+    let wallet = get_main_wallet();
+    map_ariesvcx_result(wallet.replace_did_keys_start(&did).await)
+}
+
+pub async fn rotate_verkey_apply(did: &str, temp_vk: &str) -> LibvcxResult<()> {
+    let profile = match get_main_profile() {
```
Comment (gmulhearn) [inline libvcx/src/api_lib/api_handle/wallet.rs:34]: sorry, very nitpicky, but can this just be `get_main_profile()?`

File: libvcx/src/api_lib/errors/error.rs [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+    // A2A
+    #[error("Invalid HTTP response.")]
+    InvalidHttpResponse,
+    #[error("Error Retrieving messages from API")]
+    InvalidMessages,
+
+    #[error("Common error {}", 0)]
+    Common(u32),
+    #[error("Libndy error {}", 0)]
+    LibndyError(u32),
```
Comment (gmulhearn) [inline libvcx/src/api_lib/errors/error.rs:186]: libindy?

File: libvcx/src/api_lib/errors/error.rs [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+    #[error("Unknown libindy error")]
+    UnknownLibndyError,
```
Comment (gmulhearn) [inline libvcx/src/api_lib/errors/error.rs:188]: ^

**Review by gmulhearn — CHANGES_REQUESTED** [review]
Nice refactor! looks good, just left a few comments.

All the libvcx error handling makes me wish we had something like uniffi... ;). UniFFI lets you pass `thiserror`'s thru the layer and they're automatically transformed into kotlin/swift exceptions and errors, maintaining the types and comments.

I'm tempted to try make a POC of uniffi with aries-vcx to demostrate

File: aries_vcx/src/errors/error.rs [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+    // A2A
+    #[error("Invalid HTTP response.")]
+    InvalidHttpResponse,
+    #[error("Error Retrieving messages from API")]
+    InvalidMessages,
+
+    #[error("Common error {}", 0)]
+    Common(u32),
+    #[error("Libndy error {}", 0)]
+    LibndyError(u32),
```
[the remark this answers is not in the document]
Comment (Patrik-Stas) [reply aries_vcx/src/errors/error.rs:151]: Renamed to `VdrToolsError`

File: aries_vcx/src/errors/error.rs [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+    #[error("Unknown libindy error")]
+    UnknownLibndyError,
```
[the remark this answers is not in the document]
Comment (Patrik-Stas) [reply aries_vcx/src/errors/error.rs:153]: Actually unused, removed

File: aries_vcx/src/errors/mapping_messages.rs [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+    fn from(msg_err: MessagesErrorKind) -> AriesVcxErrorKind {
+        match msg_err {
+            MessagesErrorKind::InvalidState => AriesVcxErrorKind::InvalidState,
+            MessagesErrorKind::InvalidJson => AriesVcxErrorKind::InvalidJson,
+            MessagesErrorKind::IOError => AriesVcxErrorKind::IOError,
+            MessagesErrorKind::InvalidDid => AriesVcxErrorKind::InvalidDid,
+            MessagesErrorKind::InvalidVerkey => AriesVcxErrorKind::InvalidVerkey,
+            MessagesErrorKind::InvalidUrl => AriesVcxErrorKind::InvalidUrl,
+            MessagesErrorKind::NotBase58 => AriesVcxErrorKind::NotBase58,
+            _ => AriesVcxErrorKind::UnknownLibndyError,
```
[the remark this answers is not in the document]
Comment (Patrik-Stas) [reply aries_vcx/src/errors/mapping_messages.rs:22]: Absolutely, changed to `UnknownError`

[the remark this answers is not in the document]
Comment (Patrik-Stas) [reply aries_vcx/src/errors/mapping_messages.rs:22]: Removed the `_` and added the unhandled variants - turns the only missing one was this
```
MessagesErrorKind::SerializationError => AriesVcxErrorKind::SerializationError,
```

File: aries_vcx/src/errors/mod.rs (hunk 1 above)
Comment (Patrik-Stas) [reply aries_vcx/src/errors/mod.rs:5]: oh nice! Totally missed that, moving it

File: libvcx/src/api_lib/errors/error.rs (hunk 4 above)
Comment (Patrik-Stas) [reply libvcx/src/api_lib/errors/error.rs:186]: Renamed to `VdrToolsError`

File: libvcx/src/api_lib/errors/error.rs (hunk 5 above)
Comment (Patrik-Stas) [reply libvcx/src/api_lib/errors/error.rs:188]: Removed, not used anymore

**Comment by Patrik-Stas** [comment]
> Nice refactor! looks good, just left a few comments.
  All the libvcx error handling makes me wish we had something like uniffi... ;). UniFFI lets you pass thiserror's thru the layer and they're automatically transformed into kotlin/swift exceptions and errors, maintaining the types and comments.  
  I'm tempted to try make a POC of uniffi with aries-vcx to demostrate

Thanks for review, good caches! All comments addressed. 

I'd be very cool to see uniffi in action. So far we dabbed into the `napi-rs` only. How would you go about working with uniffi? Would you create a crate on top of aries-vcx which would deal with uniffi? Or would you try to hook it directly into `aries-vcx`?

File: libvcx/src/api_lib/api_c/issuer_credential.rs (hunk 2 above)
Comment (Patrik-Stas) [reply libvcx/src/api_lib/api_c/issuer_credential.rs:558]: No worries, but behaviour was not changed. This function was previously internally calling function which was not implemented
```
pub fn generate_credential_msg(_handle: u32, _my_pw_did: &str) -> VcxResult<String> {
    Err(VcxError::from_msg(
        VcxErrorKind::ActionNotSupported,
        "Not implemented yet",
    ))
    // TODO: implement
}
```

File: libvcx/src/api_lib/api_handle/wallet.rs (hunk 3 above)
Comment (Patrik-Stas) [reply libvcx/src/api_lib/api_handle/wallet.rs:34]: Happy about nitpicking no worries :-D Fixed this anti-pattern across the board.

**Review by gmulhearn — APPROVED** [review]

**Pull request merged**